src/cunumeric/matrix/solve_template.inl

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "core/comm/coll.h"
+
+// Useful for IDEs
+#include "cunumeric/matrix/solve.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, LegateTypeCode CODE>
+struct SolveImplBody;
+
+template <LegateTypeCode CODE>
+struct support_solve : std::false_type {
+};
+template <>
+struct support_solve<LegateTypeCode::DOUBLE_LT> : std::true_type {
+};
+template <>
+struct support_solve<LegateTypeCode::FLOAT_LT> : std::true_type {
+};
+template <>
+struct support_solve<LegateTypeCode::COMPLEX64_LT> : std::true_type {
+};
+template <>
+struct support_solve<LegateTypeCode::COMPLEX128_LT> : std::true_type {
+};
+
+template <VariantKind KIND>
+struct SolveImpl {
+  template <LegateTypeCode CODE, std::enable_if_t<support_solve<CODE>::value>* = nullptr>
+  void operator()(Array& a_array, Array& b_array) const
+  {
+    using VAL = legate_type_of<CODE>;
+
+#ifdef DEBUG_CUNUMERIC
+    assert(a_array.dim() == 2);
+    assert(b_array.dim() == 1 || b_array.dim() == 2);
+#endif
+    const auto a_shape = a_array.shape<2>();
+
+    const int64_t m = a_shape.hi[0] - a_shape.lo[0] + 1;
+    const int64_t n = a_shape.hi[1] - a_shape.lo[1] + 1;
+
+#ifdef DEBUG_CUNUMERIC
+    // The Python code guarantees this property
+    assert(m == n);
+#endif
+
+    size_t a_strides[2];
+    VAL* a = a_array.read_write_accessor<VAL, 2>(a_shape).ptr(a_shape, a_strides);
```

Comment:
Can we add a debugging check here, that
```
a_strides[0] == m && a_strides[1] == 1
```
and similarly for `b_strides` below?

Reply:
https://github.com/nv-legate/cunumeric/pull/568/commits/d85b2a5ebb8769283fb1681d5e1c406194131bbf